sympy/tensor/array/sp_utils.py

```diff
@@ -0,0 +1,123 @@
+from sympy.core.sympify import _sympify
+from sympy import Tuple, Basic
+from sympy.tensor.array.ndim_array import NDimArray
+from sympy.tensor.array.sparse_ndim_array import SparseNDimArray
+from sympy.tensor.array.dense_ndim_array import DenseNDimArray
+from sympy.core.compatibility import Iterable
+
+import functools
+
+# Row-based linked list sparse matrix(LIL)
+class lil(Basic, NDimArray):
+    """
+    Create a sparse array with Row-based linked list sparse matrix(LIL) format.
+    This data structure is efficient in incremental construction of sparse arrays.
+
+    Note
+    ====
+
+    This is an experimental implementation of algorithm, the compatibility with
+    existing format needs to be verified. Only some basic operations are implemented.
+
+    Examples
+    ========
+
+    >>> from sympy.tensor.array.sp_utils import lil
+    >>> a = lil([[0, 1, 0], [2, 0, 0]])
+    >>> a._data
+    [[1], [2]]
+    >>> a._rows
+    [[1], [0]]
+    >>> a[0, 1]
+    1
+    >>> a[0, 1] = 3
+    >>> a.tolist()
+    [[0, 3, 0], [2, 0, 0]]
```

Comment:
Only some basic operations are tested. I have briefly tested addition and multiplication, it didn't work but there was not an error either. I will check it later. :-)